### PR TITLE
Show a list of users from the same geohash

### LINKED ISSUE #7: Show a list of users from the same geohash

### PR DESCRIPTION
**Pull request by flacial**
### Progress

- Made it possible for users to sign through a WebSocket request. Once they sign in, a newly unique username is generated and they're added to their corresponding geohash grid set
- We locally store the user data such as their geohash or username

### Changes in this PR

- Add a socket event that sends back all the users of a grid
- Create a folder for socket event handlers to tidy things up
- Update the client demo to listen to the `users` event that's emitted from the server and show the users as an unordered list

### CODE REVIEW

File: public/index.html [hunk 1]
```diff
@@ -10,40 +10,64 @@
...
+      }
+
+    displayUserInfo(user?.hash, user?.username)
+
+
     const socket = io({
       query: {
-        hash: userHash,
+        hash: user?.hash,
+        username: user?.username,
```
Comment (flacial) [inline public/index.html:42]: This change was made because the hash doesn't help us identify this user or other users to identify it. So, the user now locally stores their unique username.

File: websocket/handlers/signin.handler.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+  grid.add({
+    username: randomName,
+    hash,
+    status: true,
+  });
+
+  socket.emit("signin", { hash, username: randomName });
+
+  finalCallback?.({ hash, username: randomName });
```
Comment (flacial) [inline websocket/handlers/signin.handler.js]: I'm not sure if this adheres to a specific design pattern, but rather than having these side effects within this function, I decided to separate them from the core logic. I think this approach promotes better code organization and maintainability by isolating the side effects in a modular way. 

Let me know what you think :)

File: index.js [hunk 3]
```diff
@@ -22,35 +16,31 @@ app.use(express.json());
...
     }
+  }
 
-    grid.add({
-      randomName,
-    });
+  socket.on(
+    "signin",
+    handleSignin(socket, grids, (result) => {
+      const { hash } = result || {};
```
Comment (JasirZaeem) [inline index.js]: This can be `const { hash: userHash } = result || {};`

File: index.js [hunk 4]
```diff
@@ -22,35 +16,31 @@ app.use(express.json());
...
-
-    let grid = grids[hash];
+  if (userHash) {
+    // Join the user to the grid
+    socket.join(userHash);
 
-    if (!grid) {
-      grids[hash] = new Set();
-      grid = grids[hash];
+    // Send all users in the grid to the newly connected user
```
Comment (JasirZaeem) [inline index.js]: Is this comment accurate ? `io.to(userHash).emit("users", [...grids[userHash]]);` this is sending the user list to everyone in that room not just the newly connected user

File: index.js [hunk 5]
```diff
@@ -22,35 +16,31 @@ app.use(express.json());
...
+  }
 
-    grid.add({
-      randomName,
-    });
+  socket.on(
+    "signin",
+    handleSignin(socket, grids, (result) => {
+      const { hash } = result || {};
+      userHash = hash;
```
Comment (JasirZaeem) [inline index.js]: This should have been `const userHash = hash;`

File: index.js [hunk 6]
```diff
@@ -22,35 +16,31 @@ app.use(express.json());
...
 
-    socket.emit("signin", { hash, username: randomName });
-  };
+      // Join the user to the grid
+      socket.join(hash);
 
-  socket.on("signin", handleSignin);
+      // Send all users in the grid to the newly connected user
+      io.to(hash).emit("users", [...grids[userHash]]);
```
Comment (JasirZaeem) [inline index.js]: What is the significance of accessing the same value with two different names in this function ?

File: websocket/handlers/signin.handler.js (hunk 2 above)
Comment (JasirZaeem) [reply websocket/handlers/signin.handler.js]: The final callback only has two lines that are doing something, `socket.join(hash);` and `io.to(hash).emit("users", [...grids[userHash]]);` as such I think that should just be included here, right now that is adding too much indirection and making this a bit hard to follow and reason about the control flow.

File: index.js (hunk 4 above)
Comment (flacial) [reply index.js]: Oops, you're right.

File: index.js (hunk 6 above)
Comment (flacial) [reply index.js]: There's no benefit. I'll update the names real quick to eliminate any confusion.

File: index.js (hunk 5 above)
Comment (flacial) [reply index.js]: This line is related to a different feature, and it can be removed.

**Review by JasirZaeem — APPROVED** [review]
